support/nrsc5.py

```diff
@@ -1,153 +1,353 @@
+                    if component_type == ComponentType.DATA:
+                        data = SIGDataComponent(component.u.data.port, component.u.data.service_data_type,
+                                                component.u.data.type, MIMEType(component.u.data.mime))
+                        components.append(SIGComponent(component_type, component.id, None, data))
+                    component_ptr = component.next
+                evt.append(SIGService(ServiceType(service.type), service.number,
+                                      self._decode(service.name), components))
+                service_ptr = service.next
+        elif type == EventType.LOT:
+            lot = c_evt.u.lot
+            evt = LOT(lot.port, lot.lot, MIMEType(lot.mime), self._decode(lot.name), lot.data[:lot.size])
+        self.callback(type, evt)
+
+    def __init__(self, callback):
+        self._load_library()
+        self.radio = c_void_p()
+        self.callback = callback
+
+    def open(self, device_index, ppm_error):
+        result = NRSC5.libnrsc5.nrsc5_open(byref(self.radio), device_index, ppm_error)
+        if result != 0:
+            raise NRSC5Error("Failed to open RTL-SDR.")
+        self._set_callback()
+
+    def open_pipe(self):
+        result = NRSC5.libnrsc5.nrsc5_open_pipe(byref(self.radio))
+        if result != 0:
+            raise NRSC5Error("Failed to open pipe.")
+        self._set_callback()
+
+    def close(self):
+        NRSC5.libnrsc5.nrsc5_close(self.radio)
+
+    def start(self):
+        NRSC5.libnrsc5.nrsc5_start(self.radio)
+
+    def stop(self):
+        NRSC5.libnrsc5.nrsc5_stop(self.radio)
+
+    def get_frequency(self):
+        frequency = c_float()
+        NRSC5.libnrsc5.nrsc5_get_frequency(self.radio, byref(frequency))
+        return frequency.value
+
+    def set_frequency(self, freq):
+        result = NRSC5.libnrsc5.nrsc5_set_frequency(self.radio, c_float(freq))
+        if result != 0:
+            raise NRSC5Error("Failed to set frequency.")
+
+    def get_gain(self):
+        gain = c_float()
+        NRSC5.libnrsc5.nrsc5_get_gain(self.radio, byref(gain))
+        return gain.value
+
+    def set_gain(self, gain):
+        result = NRSC5.libnrsc5.nrsc5_set_gain(self.radio, c_float(gain))
+        if result != 0:
+            raise NRSC5Error("Failed to set gain.")
+
+    def set_auto_gain(self, enabled):
+        NRSC5.libnrsc5.nrsc5_set_auto_gain(self.radio, int(enabled))
+
+    def _set_callback(self):
+        def callback_closure(evt, opaque):
+            self._callback_wrapper(evt)
 
-    libnrsc5.nrsc5_stop(radio)
-    libnrsc5.nrsc5_close(radio)
+        self.callback_func = CFUNCTYPE(None, POINTER(_Event), c_void_p)(callback_closure)
+        NRSC5.libnrsc5.nrsc5_set_callback(self.radio, self.callback_func, None)
 
-    audio_queue.put(None)
-    audio_thread.join()
+    def pipe_samples(self, samples):
+        result = NRSC5.libnrsc5.nrsc5_pipe_samples(self.radio, samples, (len(samples) // 4) * 4)
```

Comment:
For completeness, is it worth adding a warning or assertion if len(samples) % 4 != 0?

Reply:
Yeah, it's probably worth making it harder to use the API incorrectly. I'll add something in another PR.